src/sisl/_core/geometry.py

```diff
@@ -1685,6 +1685,85 @@ def angle(
             return ang
         return np.degrees(ang)
 
+    def dihedral(
+        self,
+        atoms: AtomsIndex,
+        rad: bool = False,
+    ) -> Union[float, ndarray]:
+        r"""Calculate the dihedral angle defined by four atoms.
+
+        The dihehral angle is defined between 2 half-planes.
+
+        The first 3 atoms define the first plane
+        The last 3 atoms define the second.
+
+        The dihedral angle is calculated using this formula:
+
+        .. math::
+
+            \mathbf u_0 &= \mathbf r_1 - \mathbf r_0
+            \\
+            \mathbf u_1 &= \mathbf r_2 - \mathbf r_1
+            \\
+            \mathbf u_2 &= \mathbf r_3 - \mathbf r_2
+            \\
+            \phi &= \operatorname{atan2}\Big(\hat\mathbf u_0\dot
+                (\hat\mathbf u_1\times\hat\mathbf u_2),
+                (\hat\mathbf u_0\times\hat\mathbf u_1)\dot
+                (\hat\mathbf u_1\times\hat\mathbf u_2)
+                \Big)
+
+        Parameters
+        ----------
+        atoms :
+           An array of shape (4,)  or (N, 4) representing the indices of 4 atoms forming the dihedral angle
+        rad :
+           whether the returned value is in radians
+        """
+        atoms = self._sanitize_atoms(atoms)
+        ndim = atoms.ndim
+        if ndim == 1:
+            if len(atoms) != 4:
+                raise ValueError(
+                    f"{self.__class__.__name__}.dihedral requires atoms to be 4 indices"
+                )
+            atoms = [atoms]
+        elif ndim == 2:
+            if atoms.shape[1] != 4:
+                raise ValueError(
+                    f"{self.__class__.__name__}.dihedral requires atoms to be (N, 4) indices"
+                )
+        else:
+            raise ValueError(
+                f"{self.__class__.__name__}.dihedral requires atoms index of shape (4,) or (N, 4)"
+            )
+
+        angles = []
+        for idx in atoms:
+            # The 2 planes are defined by
+            #  r0, r1, r2
+            # and
+            #  r1, r2, r3
+            v = np.diff(self.xyz[idx], axis=0)
```

Comment:
Should we rename to `u` to be consistent with the presented formula?

Reply:
Yes, I have another proposition, I'll play a bit with it, but perhaps we should do it the other way around?

So one can do
```
a0 = [0, 1]
a1 = [2, 3]
a2 = [4, 5]
a3 = [6, 7]
geom.dihedral([a0, a1, a2, a3])
```
that seems a bit more natural to me, rather than the current way? What is your reasoning for the other way?

Reply:
I thought about that too. I ended up with the current order because I wanted to calculate first one angle, then decided to add another one, and it was easier to write the next four indices together in a list. For me, getting `N` angles from an `(N, 4)` input array seems a natural ordering.

Reply:
yes, but how would you use it? The internal algorithm does not matter here ;) That is trivially fixed ;)

Reply:
the return dimension would look the same, regardless.

Reply:
Sure, I don't have a strong opinion on the ordering. `(4, N)` input would be fine too, as long as the documentation is clear about it.

Reply:
hmm, I can also see that one would do this:
```python
d1 = [0, 2, 4, 6]
d2 = [1, 3, 5, 7]
geom.dihedral([d1, d2])
```
hmm, now I am in doubt, what do you think would be most natural?

Reply:
You can see how I would use it in the tests.

Reply:
ok, lets keep it as is, I will amend the calculation to be done without the loop.